storm-client/test/jvm/org/apache/storm/topology/PersistentWindowedBoltExecutorTest.java

```diff
@@ -0,0 +1,261 @@
+        // late tuple emitted
+        Mockito.verify(mockOutputCollector, Mockito.times(1))
+            .emit(stringCaptor.capture(), anchorCaptor.capture(), valuesCaptor.capture());
+        Assert.assertEquals(LATE_STREAM, stringCaptor.getValue());
+        Assert.assertEquals(Collections.singletonList(mockTuple), anchorCaptor.getValue());
+        Assert.assertEquals(new Values(mockTuple), valuesCaptor.getValue());
+    }
+
+    @Test
+    public void testActivation() throws Exception {
+        Mockito.when(mockWaterMarkEventGenerator.track(Mockito.any(GlobalStreamId.class), Mockito.anyLong())).thenReturn(true);
+        executor.initState(null);
+        executor.waterMarkEventGenerator = mockWaterMarkEventGenerator;
+
+        List<Tuple> mockTuples = getMockTuples(5);
+        mockTuples.forEach(t -> executor.execute(t));
+        // all tuples acked
+        Mockito.verify(mockOutputCollector, Mockito.times(5)).ack(tupleCaptor.capture());
+        Assert.assertArrayEquals(mockTuples.toArray(), tupleCaptor.getAllValues().toArray());
+
+        Mockito.doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                TupleWindow window = (TupleWindow) invocation.getArguments()[0];
+                // iterate the tuples
+                Assert.assertEquals(5, window.get().size());
+                // iterating multiple times should produce same events
+                Assert.assertEquals(5, window.get().size());
+                Assert.assertEquals(5, window.get().size());
+                return null;
+            }
+        }).when(mockBolt).execute(Mockito.any());
+        // trigger the window
+        long activationTs = tupleTs + 1000;
+        executor.getWindowManager().add(new WaterMarkEvent<>(activationTs));
+        executor.prePrepare(0);
+
+        // partition ids
+        Mockito.verify(mockPartitionState, Mockito.times(1)).put(stringCaptor.capture(), partitionValuesCaptor.capture());
+        Assert.assertEquals(PARTITION_KEY, stringCaptor.getAllValues().get(0));
+        Assert.assertEquals(Collections.singletonList(0L), partitionValuesCaptor.getAllValues().get(0));
+
+        // window partitions
+        Mockito.verify(mockWindowState, Mockito.times(1)).put(longCaptor.capture(), windowValuesCaptor.capture());
+        Assert.assertEquals(0L, (long) longCaptor.getAllValues().get(0));
+        Assert.assertEquals(5, windowValuesCaptor.getAllValues().get(0).size());
+        List<Event<?>> tuples = windowValuesCaptor.getAllValues().get(0)
+            .getEvents().stream().map(Event::get).collect(Collectors.toList());
+        Assert.assertArrayEquals(mockTuples.toArray(), tuples.toArray());
+
+        // window system state
+        Mockito.verify(mockSystemState, Mockito.times(2)).put(stringCaptor.capture(), systemValuesCaptor.capture());
+        Assert.assertEquals(EVICTION_STATE_KEY, stringCaptor.getAllValues().get(1));
```

Comment:
Nit: Reusing the captor like this is pretty brittle when someone needs to add extra verification to this test. Consider just declaring the captors as local variables and use a captor per verify. I don't think you gain anything by using [at]Captor for non-generic classes anyway.